Update style.css: Enable Characters Hotswap Sliding and Display More Characters

RivelleDays opened the pull request:
## Checklist:
- [X] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

- - -

- **Enable Characters Hotswap Sliding and Display More Characters**
I noticed that many users (myself included) have multiple starred characters. I've always loved this feature, so I hope it can have a visually appealing scrolling effect.

RivelleDays committed:
Title: Update style.css: Enable Characters Hotswap Sliding and Display More Characters

I noticed that many users (myself included) have multiple starred characters. I've always loved this feature, so I hope it can have a visually appealing scrolling effect.

Cohee1207 commented:
I don't see any difference compared to staging.

RivelleDays commented:
> I don't see any difference compared to staging.

https://github.com/user-attachments/assets/96bdae3c-77a0-43c4-ace6-1da1d0dd11fe

The effect should look like the video attachment. This might be more noticeable on mobile, as you need enough starred characters, and the narrower screen width makes it more apparent.

*This is my first time uploading a video on GitHub, so I’m not sure what will happen…*

Cohee1207 commented on public/style.css:
```diff
@@ -1158,9 +1158,15 @@ body.swipeAllMessages .mes:not(.last_mes) .swipes-counter {
-    flex-wrap: wrap;
-    overflow: hidden;
-    max-height: calc(var(--avatar-base-height) + 2 * var(--avatar-base-border-radius));
+    flex-wrap: nowrap;
```
`nowrap` wouldn't work here because flex-container class overrides it.
Also this would apply everywhere with avatars_inline class since it's reused for group member inline lists and tag edit lists.

To resolve, create a new style block, targeting two classes at once:
```css
.hotswap.avatars_inline {
  /* Changed styles here */
}
```

Cohee1207 commented on public/style.css:
```diff
@@ -1158,9 +1158,15 @@ body.swipeAllMessages .mes:not(.last_mes) .swipes-counter {
+    overflow-x: auto;
+    overflow-y: hidden;
+    scroll-behavior: smooth;
+    -webkit-overflow-scrolling: touch;
```
We don't usually employ prefixed properties unless strictly necessary. You can safely remove both webkit and ms prefixed props.

Cohee1207 commented:
On desktop it only works after a scroll wheel click and drag. This does not feel like an organic UX pattern.

RivelleDays committed:
Update style.css

1. New class: .hotswap.avatars_inline
2. Hidden horizontal scrollbar
3. Preserved functionality: Content remains horizontally scrollable

RivelleDays commented:
Thank you so much for reviewing this!
I realize I’m still not familiar enough with SillyTavern's CSS, and I really appreciate your help!
The CSS has been updated, and please feel free to reach out if there are any issues ; v ;

Cohee1207 commented:
I tested it and it's almost good to merge, the only pain point I have is that the scroll position resets with every hotswap update. Tell me if you want to fix that yourself (requires JavaScript to memorize and restore scroll position on render).

RivelleDays commented:
Sorry, I don’t think I’m up to this… 🥹 I only know HTML/CSS, so I’m not sure I can help much.

RivelleDays committed:
Merge branch 'SillyTavern:release' into release

Wolfsblvt reviewed (commented):
It works, and similar to Custom CSS tweaks we shared in the past.

I remembered we had a discussion about this in the past in a similar PR.
You can check it up here: #2174.

I think I am still of the same opinion, that this should not be the default state. Hotswap is for **quick** access. If you need more than three or four there, use folders, or the "Favorites" filter in your character list, or use a custom CSS.

Horizontal scrolling is not really a good design on any website.

I wouldn't be opposed if this gets merged, but yeah, I personally don't think it's a good idea.

RivelleDays commented:
Thank you for your response. I believe making this a default setting would be more intuitive. While the improved CSS works, it still unexpectedly jumps to the start, which requires further investigation.

The main reasons for my initial suggestion are:

1.	Adding a character to "Favorites" doesn’t automatically place them in Hotswap, which feels unintuitive and lacks explanation in the interface. This is especially noticeable on mobile and smaller screens.

2.	On mobile, horizontal scrolling can be effective for quick access to secondary content, particularly when avatars aren’t fully visible. Features like Instagram’s Stories demonstrate this approach as a common user behavior.

3.	Hotswap is a highly convenient tool I use frequently. While I also rely on folders and tags, most of my interactions are with characters in Hotswap. Expanding Hotswap to support folders (potentially with custom icons) would make managing multiple characters or series more efficient without resorting to time-consuming workarounds like multiple accounts.

Intuitive design should minimize user actions while providing clear functionality. I appreciate your feedback, and I agree there’s room for further discussion on whether this should be merged.

RossAscends commented:
> 1. Adding a character to "Favorites" doesn’t automatically place them in Hotswap, which feels unintuitive and lacks explanation in the interface. This is especially noticeable on mobile and smaller screens.

To clarify here, Favoriting a character places them in candidacy for being in Hotswap, but the Hotswap area is ordered according to the selected Character List sorting order. The Hotswap area will show as many characters as it can within the width or the panel, in the order of your Character List.

I think the pushback you are seeing here is that we do not see Hotswaps as serving as a duplicate horizontal character list. If you want to quickly see all of your favorited characters, just tap the Star button at the top of the Character List. 

However, I think I see your intent with this change -- using the character list requires you to 'go back' to the main list, while Hotswaps are visible even when you have a specific character open in the panel. So it's one less click/tap. 

I could see this maybe for mobile only since horizontally swiping is much better than side scrolling with a mouse.

An alternate solution might be to make the Hotswap area expand vertically to show all favorited characters when you hover the mouse inside it (but this only helps PC users).

Wolfsblvt commented:
> Adding a character to "Favorites" doesn’t automatically place them in Hotswap, which feels unintuitive and lacks explanation in the interface. This is especially noticeable on mobile and smaller screens.

It does. Like Ross said, it should automatically redraw both the character list and the hotswap. I am pretty sure that is the case. It does not sort it to the beginning, because your search order didn't change.

> On mobile, horizontal scrolling can be effective for quick access to secondary content, particularly when avatars aren’t fully visible. Features like Instagram’s Stories demonstrate this approach as a common user behavior.

That is true for mobile, yes. And I agree with Ross that this could potentially be made a feature just for mobile.
That doesn't change the fact for me though that I *feel* like this is missing the intended use case for the hotswap bar.

All use cases that you describe are basically implemented with the character list. Folders, drilldown. Scrollable however you want, etc. The hotswap is not a second character list panel with just a subset of characters that'll provide the same features.
So this:

> Expanding Hotswap to support folders (potentially with custom icons) would make managing multiple characters or series more efficient without resorting to time-consuming workarounds like multiple accounts.

Is actually the exact opposite of its intended use and of convenience, in my eyes.
Even with just one folder, you'll remove one of the very few available slots on the hotswap panel to immediately go to a char. Now you only have 3 instead 4 characters to click, etc. For the folder, you'd have multiple clicks, then drilling out again, if need be, etc. That doesn't make sense. You have the character list that has the exact functionality, if you want to search for chars, drill down to them, and all that stuff.

@RossAscends 
> An alternate solution might be to make the Hotswap area expand vertically to show all favorited characters when you hover the mouse inside it (but this only helps PC users).

I don't see this being an actual good idea. Some users might have dozens of favorite characters, and they will not fit on the screen. Also it would make the "first" hotswap chars jump vertically to a new position when you hover over them, instead of making them very fast and easy to click on. That's counter-productive again.

RivelleDays commented:
I’m sorry for closing the PR before checking for any responses—I only noticed it when I saw the email notification.

I initially thought this PR might not need to be merged into SillyTavern itself, as it could work as part of a custom CSS style. Additionally, I was unable to resolve the issue where Hotswap jumps after selecting a character avatar (thinking now, could it be related to the sorting order?).

I use Hotswap as a draggable character menu because it’s faster on mobile. SillyTavern is nearly unusable as a PWA on iPhone (severe lag and screen clipping), and the blank space at the top is actually quite convenient.

Attached is a screenshot from my iPhone 14 Pro using the default theme. I understand that SillyTavern isn’t designed for mobile use, but I still really enjoy it. As far as I know, many users in Chinese-speaking regions also prefer using it on their phones.

![image](https://github.com/user-attachments/assets/b8c6be02-5d8e-4dd5-ac64-033b0fa4844f)
![image](https://github.com/user-attachments/assets/b943eac7-7d18-45d2-8ebb-8f06f9702971)

RivelleDays commented:
@Wolfsblvt
> Horizontal scrolling is not really a good design on any website.

I just realized you’re right. I primarily use macOS, where horizontal scrolling isn’t an issue for me. However, on Windows, you need to hold the middle mouse button to drag horizontally… that’s indeed a frustrating design.

@RossAscends
> I could see this maybe for mobile only since horizontally swiping is much better than side scrolling with a mouse.

Yes, I’m considering limiting the fix to mobile devices only. I’d like to first gather more feedback on other users’ habits. So far, I’ve found that users on mobile devices seem to like it.

Cohee1207 commented:
> I was unable to resolve the issue where Hotswap jumps after selecting a character avatar (thinking now, could it be related to the sorting order?).

This is because the list updates, and is unfixable without JS changes (save scroll position before clearing the list, restore after), but even if fixed in JS, this logic won't ever be triggered without very specific custom CSS.

Wolfsblvt commented:
> SillyTavern is nearly unusable as a PWA on iPhone (severe lag and screen clipping)
Performance profiling would help. But this mostly applies to having too many elements displayed at once. Which a lot of phones have issues with. ST *redraws* UI elements very often, which might be no issue on PC, but for mobile it can tank the performance a bit.

How many characters do you have? How many of them are displayed per page? That's one of the biggest issues. Same for the WI menu, where you shouldn't display many entries on the same page.
Limit the messages displayed in chat as well, make sure that only the recent messages are shown. This is just visual, and has no impact on what is being sent to the model.

> Additionally, I was unable to resolve the issue where Hotswap jumps after selecting a character avatar (thinking now, could it be related to the sorting order?).

As Cohee said, this is related to the char list (including the hotswaps) being redrawn in many, many places. Of course for example when you select a character.
This is on purpose.
